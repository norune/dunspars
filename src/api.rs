use std::collections::HashMap;

use anyhow::{bail, Result};
use regex::Regex;

use dirs;

use rustemon::client::{CACacheManager, RustemonClient, RustemonClientBuilder};
use rustemon::games::version_group as rustemon_version;
use rustemon::moves::move_ as rustemon_moves;
use rustemon::pokemon::pokemon_species as rustemon_species;
use rustemon::pokemon::{
    ability as rustemon_ability, pokemon as rustemon_pokemon, type_ as rustemon_type,
};
use rustemon::Follow;

use rustemon::model::evolution::{
    ChainLink as RustemonEvoStep, EvolutionChain as RustemonEvoRoot,
    EvolutionDetail as RustemonEvoMethod,
};
use rustemon::model::games::VersionGroup as RustemonVersion;
use rustemon::model::moves::{Move as RustemonMove, PastMoveStatValues as RustemonPastMoveStats};
use rustemon::model::pokemon::{
    Ability as RustemonAbility, AbilityEffectChange as RustemonPastAbilityEffect,
    Pokemon as RustemonPokemon, PokemonStat as RustemonStat, PokemonType as RustemonTypeSlot,
    PokemonTypePast as RustemonPastPokemonType, Type as RustemonType,
    TypeRelations as RustemonTypeRelations, TypeRelationsPast as RustemonPastTypeRelations,
};
use rustemon::model::resource::{Effect as RustemonEffect, VerboseEffect as RustemonVerboseEffect};

use crate::pokemon::{
    Ability, EvolutionMethod, EvolutionStep, Move, PokemonData, Stats, Type, TypeChart,
};

pub struct ApiWrapper {
    client: RustemonClient,
}

impl ApiWrapper {
    pub fn try_new() -> Result<Self> {
        let mut client = RustemonClientBuilder::default();
        let mut cache_dir = if let Some(home_dir) = dirs::home_dir() {
            home_dir
        } else {
            bail!("Home directory not found")
        };
        cache_dir.push(".cache/dunspars/rustemon");

        let cache_manager = CACacheManager { path: cache_dir };
        client.with_manager(cache_manager);

        Ok(Self {
            client: client.try_build()?,
        })
    }
}

impl ApiWrapper {
    pub async fn get_pokemon(&self, pokemon: &str, game: &str) -> Result<PokemonData> {
        let RustemonPokemon {
            name,
            types,
            past_types,
            moves,
            stats,
            abilities,
            species,
            ..
        } = rustemon_pokemon::get_by_name(pokemon, &self.client).await?;
        let generation = self.get_generation(game).await?;

        let pokemon_types = self
            .match_past(generation, past_types)
            .await
            .unwrap_or(types);
        let primary_type = pokemon_types
            .iter()
            .find(|t| t.slot == 1)
            .unwrap()
            .type_
            .name
            .clone();
        let secondary_type = pokemon_types
            .iter()
            .find(|t| t.slot == 2)
            .map(|t| t.type_.name.clone());

        let mut learn_moves = HashMap::new();
        moves.iter().for_each(|mv| {
            let version_group = mv
                .version_group_details
                .iter()
                .find(|vg| vg.version_group.name == game);
            if let Some(vg) = version_group {
                learn_moves.insert(
                    mv.move_.name.clone(),
                    (vg.move_learn_method.name.clone(), vg.level_learned_at),
                );
            }
        });

        let abilities = abilities
            .iter()
            .map(|a| (a.ability.name.clone(), a.is_hidden))
            .collect::<Vec<_>>();

        Ok(PokemonData {
            name,
            primary_type,
            secondary_type,
            learn_moves,
            abilities,
            species: species.name,
            stats: extract_stats(stats),
            game: game.to_string(),
            generation,
            api: self,
        })
    }

    pub async fn get_type(&self, type_str: &str, generation: u8) -> Result<Type> {
        let RustemonType {
            name,
            damage_relations,
            past_damage_relations,
            ..
        } = rustemon_type::get_by_name(type_str, &self.client).await?;

        let relations = self
            .match_past(generation, past_damage_relations)
            .await
            .unwrap_or(damage_relations);

        let mut offense_chart = HashMap::new();
        let mut defense_chart = HashMap::new();

        relations.no_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 0.0);
        });
        relations.half_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 0.5);
        });
        relations.double_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 2.0);
        });

        relations.no_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 0.0);
        });
        relations.half_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 0.5);
        });
        relations.double_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 2.0);
        });

        Ok(Type {
            name,
            offense_chart: TypeChart::new(offense_chart),
            defense_chart: TypeChart::new(defense_chart),
            generation,
            api: self,
        })
    }

    pub async fn get_move(&self, name: &str, generation: u8) -> Result<Move> {
        let RustemonMove {
            name,
            mut accuracy,
            mut power,
            mut pp,
            damage_class,
            mut type_,
            mut effect_chance,
            effect_entries,
            effect_changes,
            past_values,
            ..
        } = rustemon_moves::get_by_name(name, &self.client).await?;

        let RustemonVerboseEffect {
            mut effect,
            mut short_effect,
            ..
        } = effect_entries
            .into_iter()
            .find(|e| e.language.name == "en")
            .unwrap_or_default();

        if let Some(past_stats) = self.match_past(generation, past_values).await {
            accuracy = past_stats.accuracy.or(accuracy);
            power = past_stats.power.or(power);
            pp = past_stats.pp.or(pp);
            effect_chance = past_stats.effect_chance.or(effect_chance);

            if let Some(t) = past_stats.type_ {
                type_ = t;
            }

            if let Some(entry) = past_stats
                .effect_entries
                .into_iter()
                .find(|e| e.language.name == "en")
            {
                effect = entry.effect;
                short_effect = entry.short_effect;
            }
        }

        if let Some(past_effects) = self.match_past(generation, effect_changes).await {
            if let Some(past_effect) = past_effects.into_iter().find(|e| e.language.name == "en") {
                effect += format!("\n\nGeneration {generation}: {}", past_effect.effect).as_str();
            }
        }

        Ok(Move {
            name,
            accuracy,
            power,
            pp,
            damage_class: damage_class.name,
            type_: type_.name,
            effect_chance,
            effect,
            short_effect,
            generation,
            api: self,
        })
    }

    pub async fn get_ability(&self, name: &str, generation: u8) -> Result<Ability> {
        let RustemonAbility {
            name,
            effect_entries,
            effect_changes,
            ..
        } = rustemon_ability::get_by_name(name, &self.client).await?;

        let RustemonVerboseEffect {
            mut effect,
            short_effect,
            ..
        } = effect_entries
            .into_iter()
            .find(|e| e.language.name == "en")
            .unwrap_or_default();

        if let Some(past_effects) = self.match_past(generation, effect_changes).await {
            if let Some(past_effect) = past_effects.into_iter().find(|e| e.language.name == "en") {
                effect += format!("\n\nGeneration {generation}: {}", past_effect.effect).as_str();
            }
        }

        Ok(Ability {
            name,
            effect,
            short_effect,
            generation,
            api: self,
        })
    }

    pub async fn get_generation(&self, game: &str) -> Result<u8> {
        let RustemonVersion { generation, .. } =
            rustemon_version::get_by_name(game, &self.client).await?;

        extract_gen_from_url(&generation.url)
    }

    pub async fn get_evolution_steps(&self, species: &str) -> Result<EvolutionStep> {
        let RustemonEvoRoot { chain, .. } = rustemon_species::get_by_name(species, &self.client)
            .await?
            .evolution_chain
            .unwrap()
            .follow(&self.client)
            .await?;
        let evolution_step = traverse_chain(chain);

        Ok(evolution_step)
    }

    pub async fn get_all_moves(&self) -> Result<Vec<String>> {
        Ok(rustemon_moves::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    pub async fn get_all_abilities(&self) -> Result<Vec<String>> {
        Ok(rustemon_ability::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }
    pub async fn get_all_games(&self) -> Result<Vec<String>> {
        Ok(rustemon_version::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    pub async fn get_all_types(&self) -> Result<Vec<String>> {
        Ok(rustemon_type::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }
    pub async fn get_all_pokemon(&self) -> Result<Vec<String>> {
        Ok(rustemon_pokemon::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    async fn match_past<T: Past<U>, U>(&self, current_generation: u8, pasts: Vec<T>) -> Option<U> {
        let mut oldest_value = None;
        let mut oldest_generation = 255;

        for past in pasts {
            let past_generation = past.generation(self).await;
            if current_generation <= past_generation && past_generation <= oldest_generation {
                oldest_value = Some(past.value());
                oldest_generation = past_generation;
            }
        }

        oldest_value
    }
}

fn traverse_chain(chain_link: RustemonEvoStep) -> EvolutionStep {
    let evolution_methods = chain_link
        .evolution_details
        .into_iter()
        .map(convert_to_evolution_method)
        .collect();

    if !chain_link.evolves_to.is_empty() {
        let evolves_to = chain_link
            .evolves_to
            .into_iter()
            .map(traverse_chain)
            .collect();

        EvolutionStep::new(chain_link.species.name, evolution_methods, evolves_to)
    } else {
        EvolutionStep::new(chain_link.species.name, evolution_methods, vec![])
    }
}

fn convert_to_evolution_method(evolution: RustemonEvoMethod) -> EvolutionMethod {
    let mut method = EvolutionMethod::new(evolution.trigger.name);
    if let Some(item) = evolution.item {
        method = method.item(item.name);
    }
    if let Some(gender) = evolution.gender {
        method = method.gender(gender);
    }
    if let Some(held_item) = evolution.held_item {
        method = method.held_item(held_item.name);
    }
    if let Some(known_move) = evolution.known_move {
        method = method.known_move(known_move.name);
    }
    if let Some(known_move_type) = evolution.known_move_type {
        method = method.known_move_type(known_move_type.name);
    }
    if let Some(location) = evolution.location {
        method = method.location(location.name);
    }
    if let Some(min_level) = evolution.min_level {
        method = method.min_level(min_level);
    }
    if let Some(min_happiness) = evolution.min_happiness {
        method = method.min_happiness(min_happiness);
    }
    if let Some(min_beauty) = evolution.min_beauty {
        method = method.min_beauty(min_beauty);
    }
    if let Some(min_affection) = evolution.min_affection {
        method = method.min_affection(min_affection);
    }
    if let Some(party_species) = evolution.party_species {
        method = method.party_species(party_species.name);
    }
    if let Some(party_type) = evolution.party_type {
        method = method.party_type(party_type.name);
    }
    if let Some(relative_physical_stats) = evolution.relative_physical_stats {
        method = method.relative_physical_stats(relative_physical_stats);
    }
    if let Some(trade_species) = evolution.trade_species {
        method = method.trade_species(trade_species.name);
    }
    if evolution.needs_overworld_rain {
        method = method.needs_overworld_rain(true);
    }
    if evolution.turn_upside_down {
        method = method.turn_upside_down(true);
    }
    if !evolution.time_of_day.is_empty() {
        method = method.time_of_day(evolution.time_of_day);
    }

    method
}

fn extract_gen_from_url(url: &str) -> Result<u8> {
    let gen_url_regex = Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap();
    let caps = gen_url_regex.captures(url).unwrap();
    let generation = caps["gen"].parse::<u8>()?;
    Ok(generation)
}

fn extract_stats(stats_vec: Vec<RustemonStat>) -> Stats {
    let mut stats = Stats::default();

    for RustemonStat {
        stat, base_stat, ..
    } in stats_vec
    {
        match stat.name.as_str() {
            "hp" => stats.hp = base_stat,
            "attack" => stats.attack = base_stat,
            "defense" => stats.defense = base_stat,
            "special-attack" => stats.special_attack = base_stat,
            "special-defense" => stats.special_defense = base_stat,
            "speed" => stats.speed = base_stat,
            _ => (),
        }
    }

    stats
}

trait Past<T> {
    async fn generation(&self, api: &ApiWrapper) -> u8;
    fn value(self) -> T;
}

impl Past<Vec<RustemonTypeSlot>> for RustemonPastPokemonType {
    async fn generation(&self, _api: &ApiWrapper) -> u8 {
        extract_gen_from_url(&self.generation.url).unwrap()
    }

    fn value(self) -> Vec<RustemonTypeSlot> {
        self.types
    }
}

impl Past<RustemonTypeRelations> for RustemonPastTypeRelations {
    async fn generation(&self, _api: &ApiWrapper) -> u8 {
        extract_gen_from_url(&self.generation.url).unwrap()
    }

    fn value(self) -> RustemonTypeRelations {
        self.damage_relations
    }
}

impl Past<RustemonPastMoveStats> for RustemonPastMoveStats {
    async fn generation(&self, api: &ApiWrapper) -> u8 {
        api.get_generation(&self.version_group.name).await.unwrap() - 1
    }

    fn value(self) -> RustemonPastMoveStats {
        self
    }
}

impl Past<Vec<RustemonEffect>> for RustemonPastAbilityEffect {
    async fn generation(&self, api: &ApiWrapper) -> u8 {
        api.get_generation(&self.version_group.name).await.unwrap() - 1
    }

    fn value(self) -> Vec<RustemonEffect> {
        self.effect_entries
    }
}
