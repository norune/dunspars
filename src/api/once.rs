use crate::resource::GameResource;
use regex::Regex;
use std::sync::OnceLock;

pub fn game_resource() -> &'static GameResource {
    static GAME_RESOURCE: OnceLock<GameResource> = OnceLock::new();
    GAME_RESOURCE.get_or_init(|| GameResource::try_new().unwrap())
}

pub fn gen_url_regex() -> &'static Regex {
    static GEN_URL_REGEX: OnceLock<Regex> = OnceLock::new();
    GEN_URL_REGEX.get_or_init(|| {
        // Pok√©API keeps generation names in Roman numerals.
        // Might be quicker to just take it from resource urls via regex instead.
        // Regex compilation is expensive, so we're compiling it just once here.
        Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap()
    })
}
