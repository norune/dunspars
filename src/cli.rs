mod display;
mod utils;

use anyhow::{anyhow, Result};
use clap::{Parser, Subcommand};
use indoc::{formatdoc, printdoc};

use crate::data::api::resource::{
    AbilityResource, GameResource, GetGeneration, MoveResource, PokemonResource, Resource,
    TypeResource,
};
use crate::data::api::ApiWrapper;
use crate::data::{Ability, Move, Pokemon, PokemonData, Type};
use display::*;

const VERSION: &str = env!("DUNSPARS_VERSION");

#[derive(Parser)]
#[command(author, version = VERSION, about, long_about = None)]
struct Cli {
    #[command(subcommand)]
    command: Commands,
    /// Sets the mainline Pokémon game the output will be based on
    #[clap(short, long, global = true)]
    game: Option<String>,
    /// Force output to include colors
    #[clap(long, action = clap::ArgAction::SetTrue, global = true)]
    color: bool,
    /// Force output to exclude colors
    #[clap(long, action = clap::ArgAction::SetTrue, global = true)]
    no_color: bool,
}

#[derive(Subcommand)]
enum Commands {
    /// Prints general data about a Pokémon
    Pokemon {
        /// Name of the Pokémon
        pokemon: String,
        /// Display all move data the Pokémon is capable of learning
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        moves: bool,
        /// Display the Pokémon evolutionary line
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        evolution: bool,
    },
    /// Prints matchup data between Pokémon
    Match {
        /// Names of the defending Pokémon; max 6
        #[arg(required = true, num_args = 1..=6)]
        defenders: Vec<String>,
        /// Name of the attacking Pokémon
        attacker: String,
        /// Display only moves that match the user's type
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        stab_only: bool,
        /// Display verbose output
        #[arg(short, long, action = clap::ArgAction::SetTrue)]
        verbose: bool,
    },
    /// Prints type coverage based on the provided Pokémon
    Coverage {
        /// Names of Pokémon; max 6
        #[arg(required = true, num_args = 1..=6)]
        pokemon: Vec<String>,
    },
    /// Prints data about a Pokémon type
    Type {
        /// Name of the type
        type_: String,
    },
    /// Prints data about a Pokémon move
    Move {
        /// Name of the move
        move_: String,
    },
    /// Prints data about a Pokémon ability
    Ability {
        /// Name of the ability
        ability: String,
    },
    /// Prints all possible names from a Resource such as Pokémon, Moves, etc
    Resource {
        /// Name of the resource
        #[arg(value_enum)]
        resource: ResourceArgs,
        /// Value to be printed in between values. Defaults to newline
        #[arg(short, long)]
        delimiter: Option<String>,
    },
    /// Actions regarding the program's cache
    Cache {
        /// Action to undertake
        #[arg(value_enum)]
        action: CacheAction,
    },
}

#[derive(Clone, clap::ValueEnum)]
enum ResourceArgs {
    Pokemon,
    Moves,
    Abilities,
    Games,
    Types,
}

#[derive(Clone, clap::ValueEnum)]
enum CacheAction {
    Clear,
}

pub async fn run() -> Result<()> {
    let cli = Cli::parse();
    let api = ApiWrapper::try_new().await?;

    let mut config_builder = ConfigBuilder::new(&api.game_resource);
    if let Some(game) = cli.game {
        config_builder = config_builder.game(game);
    }
    if cli.color {
        config_builder = config_builder.color_enabled(true);
    }
    if cli.no_color {
        config_builder = config_builder.color_enabled(false);
    }
    let config = config_builder.build()?;

    let program = Program::new(config, api);

    match cli.command {
        Commands::Pokemon {
            pokemon,
            moves,
            evolution,
        } => print!("{}", program.run_pokemon(pokemon, moves, evolution).await?),
        Commands::Type { type_ } => print!("{}", program.run_type(type_).await?),
        Commands::Move { move_ } => print!("{}", program.run_move(move_).await?),
        Commands::Ability { ability } => print!("{}", program.run_ability(ability).await?),
        Commands::Match {
            defenders,
            attacker,
            stab_only,
            verbose,
        } => print!(
            "{}",
            program
                .run_match(defenders, attacker, verbose, stab_only)
                .await?
        ),
        Commands::Coverage { pokemon } => print!("{}", program.run_coverage(pokemon).await?),
        Commands::Resource {
            resource,
            delimiter,
        } => program.run_resource(resource, delimiter).await?,
        Commands::Cache { action } => program.run_cache(action).await?,
    }

    Ok(())
}

struct ConfigBuilder<'a> {
    game: Option<String>,
    color_enabled: Option<bool>,
    game_resource: &'a GameResource,
}

impl<'a> ConfigBuilder<'a> {
    pub fn new(game_resource: &'a GameResource) -> Self {
        ConfigBuilder {
            game_resource,
            game: None,
            color_enabled: None,
        }
    }

    pub fn game(mut self, game: String) -> Self {
        self.game = Some(game);
        self
    }

    pub fn color_enabled(mut self, color_enabled: bool) -> Self {
        self.color_enabled = Some(color_enabled);
        self
    }

    pub fn build(self) -> Result<Config> {
        let game = match self.game {
            Some(game) => self.game_resource.validate(&game)?,
            None => self
                .get_latest_game()
                .ok_or(anyhow!("Cannot find the latest game"))?,
        };

        let generation = self.game_resource.get_gen(&game);
        let color_enabled = self.color_enabled.unwrap_or(utils::is_color_enabled());

        Ok(Config {
            game,
            generation,
            color_enabled,
        })
    }

    fn get_latest_game(&self) -> Option<String> {
        self.game_resource.resource().last().map(|g| g.to_string())
    }
}

struct Config {
    game: String,
    generation: u8,
    color_enabled: bool,
}

struct Program {
    config: Config,
    api: ApiWrapper,
}

impl Program {
    pub fn new(config: Config, api: ApiWrapper) -> Self {
        Self { config, api }
    }

    async fn run_pokemon(&self, name: String, moves: bool, evolution: bool) -> Result<String> {
        let resource = PokemonResource::try_new(&self.api.client).await?;
        let pokemon_name = resource.validate(&name)?;

        let pokemon = PokemonData::from_name(&self.api, &pokemon_name, &self.config.game).await?;
        let pokemon_display = DisplayComponent::new(&pokemon, self.config.color_enabled);

        let defense_chart = pokemon.get_defense_chart().await?;
        let defense_chart_ctx = TypeChartComponent {
            type_chart: &defense_chart,
        };
        let type_chart_display =
            DisplayComponent::new(defense_chart_ctx, self.config.color_enabled);

        let mut output = formatdoc! {
            "
            {pokemon_display}

            {type_chart_display}
            "
        };

        if evolution {
            let evolution_step = pokemon.get_evolution_steps().await?;
            let evolution_step_display =
                DisplayComponent::new(&evolution_step, self.config.color_enabled);
            output += formatdoc! {
                "

                {evolution_step_display}
                "
            }
            .as_str();
        }

        if moves {
            let moves = pokemon.get_moves().await?;
            let move_list_context = MoveListComponent {
                move_list: &moves,
                pokemon: &pokemon,
            };
            let move_list_display =
                DisplayComponent::new(move_list_context, self.config.color_enabled);

            output += formatdoc! {
                "

                {move_list_display}
                "
            }
            .as_str();
        }

        Ok(output)
    }

    async fn run_type(&self, name: String) -> Result<String> {
        let resource = TypeResource::try_new(&self.api.client).await?;
        let type_name = resource.validate(&name)?;

        let Type {
            offense_chart,
            defense_chart,
            ..
        } = Type::from_name(&self.api, &type_name, self.config.generation).await?;

        let offense_chart_ctx = TypeChartComponent {
            type_chart: &offense_chart,
        };
        let offense_chart_display =
            DisplayComponent::new(offense_chart_ctx, self.config.color_enabled);

        let defense_chart_ctx = TypeChartComponent {
            type_chart: &defense_chart,
        };
        let defense_chart_display =
            DisplayComponent::new(defense_chart_ctx, self.config.color_enabled);

        let output = formatdoc! {
            "
            {offense_chart_display}

            {defense_chart_display}
            "
        };

        Ok(output)
    }

    async fn run_match(
        &self,
        defender_names: Vec<String>,
        attacker_name: String,
        verbose: bool,
        stab_only: bool,
    ) -> Result<String> {
        let resource = PokemonResource::try_new(&self.api.client).await?;

        let attacker_name = resource.validate(&attacker_name)?;
        let attacker_data =
            PokemonData::from_name(&self.api, &attacker_name, &self.config.game).await?;
        let attacker_moves = attacker_data.get_moves().await?;
        let attacker_chart = attacker_data.get_defense_chart().await?;
        let attacker = Pokemon::new(attacker_data, attacker_chart, attacker_moves);

        let mut defenders = vec![];

        for defender_name in defender_names {
            let defender_name = resource.validate(&defender_name)?;
            let defender_data =
                PokemonData::from_name(&self.api, &defender_name, &self.config.game).await?;
            let defender_moves = defender_data.get_moves().await?;
            let defender_chart = defender_data.get_defense_chart().await?;
            let defender = Pokemon::new(defender_data, defender_chart, defender_moves);

            defenders.push(defender);
        }

        let mut output = String::from("");
        for defender in defenders {
            let match_context = MatchComponent {
                defender: &defender,
                attacker: &attacker,
                verbose,
                stab_only,
            };
            let match_display = DisplayComponent::new(match_context, self.config.color_enabled);

            output += formatdoc! {
                "
                {match_display}


                "
            }
            .as_str();
        }

        Ok(output)
    }

    async fn run_coverage(&self, names: Vec<String>) -> Result<String> {
        let resource = PokemonResource::try_new(&self.api.client).await?;
        let mut pokemon = vec![];

        for name in names {
            let name = resource.validate(&name)?;
            let data = PokemonData::from_name(&self.api, &name, &self.config.game).await?;
            let moves = data.get_moves().await?;
            let chart = data.get_defense_chart().await?;

            let mon = Pokemon::new(data, chart, moves);
            pokemon.push(mon);
        }

        let coverage_ctx = CoverageComponent::try_new(&pokemon).await?;
        let coverage_display = DisplayComponent::new(coverage_ctx, self.config.color_enabled);

        Ok(formatdoc! {
            "
            {coverage_display}
            "
        })
    }

    async fn run_move(&self, name: String) -> Result<String> {
        let resource = MoveResource::try_new(&self.api.client).await?;
        let move_name = resource.validate(&name)?;

        let move_ = Move::from_name(&self.api, &move_name, self.config.generation).await?;
        let move_display = DisplayComponent::new(&move_, self.config.color_enabled);

        let output = formatdoc! {
            "
            {move_display}
            "
        };

        Ok(output)
    }

    async fn run_ability(&self, name: String) -> Result<String> {
        let resource = AbilityResource::try_new(&self.api.client).await?;
        let ability_name = resource.validate(&name)?;

        let ability = Ability::from_name(&self.api, &ability_name, self.config.generation).await?;
        let ability_display = DisplayComponent::new(&ability, self.config.color_enabled);

        let output = formatdoc! {
            "
            {ability_display}
            "
        };

        Ok(output)
    }

    async fn run_resource(&self, resource: ResourceArgs, delimiter: Option<String>) -> Result<()> {
        let delimiter = delimiter.unwrap_or("\n".to_string());
        let resource = match resource {
            ResourceArgs::Pokemon => PokemonResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Moves => MoveResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Abilities => AbilityResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Types => TypeResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
            ResourceArgs::Games => GameResource::try_new(&self.api.client)
                .await?
                .resource()
                .join(&delimiter),
        };

        printdoc! {
            "
            {resource}
            "
        };

        Ok(())
    }

    async fn run_cache(&self, action: CacheAction) -> Result<()> {
        match action {
            CacheAction::Clear => self.api.clear_cache().await,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    async fn setup_program(game: &str) -> Program {
        let api = ApiWrapper::try_new().await.unwrap();
        let config = ConfigBuilder::new(&api.game_resource)
            .game(String::from(game))
            .color_enabled(false)
            .build()
            .unwrap();
        Program::new(config, api)
    }

    #[tokio::test]
    async fn run_pokemon() {
        let program = setup_program("scarlet-violet").await;

        let output = program
            .run_pokemon(String::from("ceruledge"), false, false)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon ceruledge --game scarlet-violet",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_pokemon_evolution() {
        let program = setup_program("sword-shield").await;

        let cascoon = program
            .run_pokemon(String::from("cascoon"), false, true)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon cascoon --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(cascoon);
        });

        let politoed = program
            .run_pokemon(String::from("politoed"), false, true)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon politoed --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(politoed);
        });

        let applin = program
            .run_pokemon(String::from("applin"), false, true)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon applin --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(applin);
        });
    }

    #[tokio::test]
    async fn run_pokemon_moves() {
        let program = setup_program("scarlet-violet").await;

        let output = program
            .run_pokemon(String::from("blaziken"), true, false)
            .await
            .unwrap();

        insta::with_settings!({
            description => "pokemon blaziken --moves --game scarlet-violet",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_type() {
        let program = setup_program("platinum").await;

        let output = program.run_type(String::from("ice")).await.unwrap();

        insta::with_settings!({
            description => "type ice --game platinum",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_move() {
        let program = setup_program("sun-moon").await;

        let output = program.run_move(String::from("brick-break")).await.unwrap();

        insta::with_settings!({
            description => "move brick-break --game sun-moon",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_ability() {
        let program = setup_program("black-white").await;

        let output = program
            .run_ability(String::from("intimidate"))
            .await
            .unwrap();

        insta::with_settings!({
            description => "ability intimidate --game black-white",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_match() {
        let program = setup_program("x-y").await;
        let defenders = vec![String::from("golem"), String::from("pachirisu")];
        let attacker = String::from("lapras");

        let non_verbose = program
            .run_match(defenders.clone(), attacker.clone(), false, false)
            .await
            .unwrap();

        let stab_only = program
            .run_match(defenders.clone(), attacker.clone(), false, true)
            .await
            .unwrap();

        let verbose = program
            .run_match(defenders.clone(), attacker.clone(), true, false)
            .await
            .unwrap();

        insta::with_settings!({
            description => "match golem pachirisu lapras --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(non_verbose);
        });

        insta::with_settings!({
            description => "match golem pachirisu lapras --stab-only --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(stab_only);
        });

        insta::with_settings!({
            description => "match golem pachirisu lapras --verbose --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(verbose);
        });
    }

    #[tokio::test]
    async fn run_coverage() {
        let program = setup_program("the-indigo-disk").await;
        let team = vec![
            String::from("flamigo"),
            String::from("cramorant"),
            String::from("ribombee"),
            String::from("ogerpon-cornerstone-mask"),
            String::from("dudunsparce"),
            String::from("sinistcha"),
        ];

        let output = program.run_coverage(team).await.unwrap();

        insta::with_settings!({
            description => "coverage flamigo cramorant ribombee ogerpon-cornerstone-mask dudunsparce sinistcha --game the-indigo-disk",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }
}
