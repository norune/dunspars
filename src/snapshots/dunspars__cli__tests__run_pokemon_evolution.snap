---
source: src/cli.rs
description: pokemon cascoon --evolution --game sword-shield
---
cascoon bug 
shed-skin
hp    atk   def   satk  sdef  spd   total
50    35    55    25    25    15    205   
gen-8

bug defense
double: fire flying rock
neutral: bug dark dragon electric fairy ghost ice normal poison psychic steel water
half: fighting grass ground

evolution
wurmple 
  silcoon level-up level-7
    beautifly level-up level-10
  cascoon level-up level-7
    dustox level-up level-10
